src/persist-client/src/dyn_cfg.rs

```diff
@@ -0,0 +1,409 @@
+            T::set(dst, T::get(src))
+        }
+
+        for (name, val) in self.updates.iter() {
+            let Some(config) = set.configs.get(name) else {
+                error!("config update {} {:?} not known set: {:?}", name, val, set);
+                continue;
+            };
+            let val = match (val.clone()).into_rust() {
+                Ok(x) => x,
+                Err(err) => {
+                    error!("config update {} decode error: {}", name, err);
+                    continue;
+                }
+            };
+            match (&val, &config.val) {
+                (ConfigVal::Bool(src), ConfigVal::Bool(dst)) => copy::<bool>(src, dst),
+                (ConfigVal::Usize(src), ConfigVal::Usize(dst)) => copy::<usize>(src, dst),
+                (ConfigVal::String(src), ConfigVal::String(dst)) => copy::<String>(src, dst),
+                (src, dst) => error!(
+                    "config update {} type mismatch: {:?} vs {:?}",
+                    name, src, dst
+                ),
+            }
+        }
+    }
+}
+
+mod impls {
+    use std::sync::atomic::AtomicBool;
+    use std::sync::atomic::{AtomicU64, Ordering::SeqCst};
+    use std::sync::{Arc, RwLock};
+
+    use mz_ore::cast::CastFrom;
+    use mz_proto::{RustType, TryFromProtoError};
+
+    use crate::dyn_cfg::{Config, ConfigSet, ConfigType, ConfigVal, ProtoConfigVal};
+
+    pub trait Sealed {}
+
+    impl ConfigType for bool {
+        type Shared = AtomicBool;
+
+        fn shared<'a>(config: &Config<Self>, vals: &'a ConfigSet) -> Option<&'a Arc<Self::Shared>> {
+            let entry = vals.configs.get(config.name)?;
+            match entry.val() {
+                ConfigVal::Bool(x) => Some(x),
+                x => panic!("expected bool value got {:?}", x),
+            }
+        }
+        fn to_val(val: &Self) -> ConfigVal {
+            ConfigVal::Bool(Arc::new((*val).into()))
+        }
+        fn set(x: &Self::Shared, val: Self) {
+            x.store(val, SeqCst);
+        }
+        fn get(x: &Self::Shared) -> Self {
+            x.load(SeqCst)
+        }
+    }
+    impl Sealed for bool {}
+
+    impl ConfigType for usize {
+        type Shared = AtomicU64;
+
+        fn shared<'a>(config: &Config<Self>, vals: &'a ConfigSet) -> Option<&'a Arc<Self::Shared>> {
+            let entry = vals.configs.get(config.name)?;
+            match entry.val() {
+                ConfigVal::Usize(x) => Some(x),
+                x => panic!("expected usize value got {:?}", x),
+            }
+        }
+        fn to_val(val: &Self) -> ConfigVal {
+            ConfigVal::Usize(Arc::new(u64::cast_from(*val).into()))
+        }
+        fn set(x: &Self::Shared, val: Self) {
+            x.store(u64::cast_from(val), SeqCst);
+        }
+        fn get(x: &Self::Shared) -> Self {
+            usize::cast_from(x.load(SeqCst))
+        }
+    }
+    impl Sealed for usize {}
+
+    impl ConfigType for String {
```

Comment:
I do think it would be fine to punt on `String` support... but this does feel like a weird interim state, where it's sort-of supported, but anyone who tries to actually declare a `String` config will be unable to do so without reworking the core trait.

If you'd rather leave it for followup work, leaving a warning comment and/or dropping this impl also seem fine to me.

Reply:
Cool thanks! And yeah, definitely easy enough to tweak later if we need to...